enhancements/namespace-auto-pruning.md

```diff
@@ -0,0 +1,170 @@
+* Allowing policy configurations at the repository level
+
+## Proposal
+
+The details in this proposal is only for namespace auto-pruning policies but designed in a way to support repository auto-pruning policies in the future. This proposal is structured in three parts: how policies are managed, how the policies are structured in the DB, and how the policies will be executed via the auto-prune worker.
+
+### Managing Policies
+
+A user is able to configure auto-pruning within the namespace settings. To start, there will be two auto-pruning options:
+* Prune by number of tags, provided the number of tags
+* Prune tags older than date, provided the number of days 
+
+For this implementation users will only be able to configure one policy per namespace. To interface with this new UI there will be a new set of CRUD endpoints for managing auto-pruning policies. The endpoints returning lists will be paginated.
+
+**Create policy**
+```
+POST /api/v1/organization/{organization}/autoprunepolicy
+By date:
+{
+    pruneBy: "creation_date",
+    olderThan: "2w"  
+}
+By number of tags:
+{
+    pruneBy: "number_of_tags",
+    numTags: 100  
+}
+```
+
+**Update policy**
+```
+PATCH /api/v1/organization/{organization}/autoprunepolicy/{policyid}
+{
+    fieldToBeUpdated: value,
+}
+```
+
+**List policies**
+```
+GET /api/v1/organization/{organization}/autoprunepolicy
+```
+
+**Get policy**
+```
+GET /api/v1/organization/{organization}/autoprunepolicy/{policyid}
+```
+
+**Delete policy**
+```
+DELETE /api/v1/organization/{organization}/autoprunepolicy/{policyid}
+```
+
+### Database structure
+Two tables will be required.
+
+**namespaceautoprunepolicy**
+
+This table holds the policy configuration for a single namespace. For this proposal there will be only one entry per namespace but there will be support for multiple rows per namespace_id. The config field will hold the policy details such as `{pruneBy: "creation_date", olderThan: "2w"}` or `{pruneBy: "number_of_tags",  numTags: 100}`.
+
+| Field | Datatype | Attributes | Description|
+| --- | ----------- | ----------- | ----------- |
+| uuid | character varying(255) | Unique, not null, indexed | Unique identifier for this policy |
+| namespace_id | integer | FK, not null | Namespace that the policy falls under |
+| config | text | JSON, not null | Policy configuration |
+
+**autoprunetask**
+
+This table registers tasks to be executed by the auto-prune worker. Tasks are executed within the context of a single namespace. Only one task per namespace will ever exist.
+
+| Field | Datatype | Attributes | Description |
+| --- | ----------- | ----------- | ----------- |
+| namespace_id | integer | FK, not null | Namespace that this task belongs too |
+| last_ran_ms | bigint | nullable, indexed | Last time the worker executed the policies for this namespace |
+| is_running | boolean | default false | Policies for this namespace are currently being executed |
```

Comment:
`Is_running` will incur an extra write on every execution, so I'd suggest against it.
The application should be able to work out if the task is running if the row is locked (e.g. if `SELECT 1 FROM autoprunetask WHERE namespace_id = $1 SKIP LOCKED;` returns nothing).

This table is also missing reporting for the status of the previous run; I'd suggest adding a `TEXT` column for holding an error return (NULL meaning no error), or a `TEXT` and a `BOOLEAN` for "status" and "errored".

Reply:
> The application should be able to work out if the task is running if the row is locked

When exactly does the row get locked in the first place? At the start of the transaction? When exactly does that lock get released?

> This table is also missing reporting for the status of the previous run; I'd suggest adding a TEXT column for holding an error return

Agreed, good idea

Reply:
The query feeding the worker has some row-level locking, right? The way this is done with PostgresQL is `SELECT ... FROM ... WHERE ... LIMIT ... FOR UPDATE SKIP LOCKED;` (specific strength of lock depends).

Reply:
ah okay i see now. Yeah that's a much better method. Looks like there's methods already in quay supporting that. I'll use those instead

Reply:
Quay uses redis for locks in parts of the code like GC for example

Reply:
it looks like the queue worker uses the `for update` @hdonnay is describing. Would we want to use that instead to remove the dependency on redis?

Reply:
@hdonnay I added the `FOR UPDATE SKIP LOCKED` method to the document. This has benefits over the redis lock where we don't have to make multiple requests to find an unlocked row.